### PR TITLE
add support for RawRequest

### PR DESCRIPTION
**Pull request by xavdid-stripe**
### Why?

As the v2 API is releasing soon, we want Go to be able to make requests against the new endpoints despite not having generated types.

### What

This is a copy/paste of https://github.com/stripe/stripe-go/pull/1648 with a few key differences:

- updated README with a more modern example
- changed the `APIMode` constants from `"preview"` and `"standard"` to `"v1"` and `"v2"`
- ❗inferred API mode from path prefix instead of letting user specify it
- removed some assertions in tests that cared about the API version

Because we no longer needed to supply default values for API mode, I could remove the function that handled that.  I also adjusted tests that assumed you could set `content-type` and `v1` / `v2` independently (given that all v1 requests are form encoded and all v2 request are json).

### Manual Testing

1. Create a new folder, henceforth `myfolder`
2. In `myfolder/go.mod`, paste:

```go
module xavd.id/test/v2

go 1.22.4

require github.com/stripe/stripe-go/v79 v79.0.0
// may need to adjust this relative path for your filesystem:
replace github.com/stripe/stripe-go/v79 v79.0.0 => ../../../stripe-go
```

3. In `myfolder/main.go` paste:

```go
package main

import (
	"encoding/json"
	"fmt"

	"github.com/stripe/stripe-go/v79"
	"github.com/stripe/stripe-go/v79/form"
	"github.com/stripe/stripe-go/v79/rawrequest"
)

func do_thing() error {
	stripe.Key = "sk_test_..."

	payload := map[string]interface{}{
		"event_name": "hotdogs_eaten", // FIXME
		"payload": map[string]string{
			"value":              "123", 
			"stripe_customer_id": "cus_Quq8itmW58RMet", // FIXME
		},
	}

	// for a v2 request, json encode the payload
	body, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	v2_resp, err := rawrequest.Post("/v2/billing/meter_events", string(body), nil)
	if err != nil {
		return err
	}

	var v2_response map[string]interface{}
	err = json.Unmarshal(v2_resp.RawJSON, &v2_response)
	if err != nil {
		return err
	}
	fmt.Printf("%#v\n", v2_response)

	// for a v1 request, form encode the payload
	formValues := &form.Values{}
	form.AppendTo(formValues, payload)
	content := formValues.Encode()

	v1_resp, err := rawrequest.Post("/v1/billing/meter_events", content, nil)
	if err != nil {
		return err
	}

	var v1_response map[string]interface{}
	err = json.Unmarshal(v1_resp.RawJSON, &v1_response)
	if err != nil {
		return err
	}
	fmt.Printf("%#v\n", v1_response)

	return nil
}

func main() {
	err := do_thing()
	if err != nil {
		fmt.Printf("Error: %s\n", err)
	}
}
```

Running that (with valid customer id and event name) should print 2 response bodies

### See Also

- [DEVSDK-2187](https://jira.corp.stripe.com/browse/DEVSDK-2187)

### CODE REVIEW

File: README.md
Comment (xavdid-stripe) [inline README.md:2]: this got prettier-ed, but the only actual changes are the section I added

File: README.md
Comment (xavdid-stripe) [inline README.md:565]: I may also add a simplified example here tomorrow before merge, but this example also works

File: params.go
Comment (xavdid-stripe) [inline params.go:177]: changed from

```go
var PreviewAPIMode APIMode = "preview"
var StandardAPIMode APIMode = "standard"
```

File: stripe_test.go
Comment (xavdid-stripe) [inline stripe_test.go:1324]: removed for now, but maybe useful later

File: api_version.go
Comment (xavdid-stripe) [inline api_version.go:10]: I changed this to get my local example working, but am not sure it should be included in this PR

Comment (ramya-stripe) [reply api_version.go:10]: change the base of the PR from master to sdk-release/next-major - that branch has this change

File: params.go
Comment (ramya-stripe) [reply params.go:177]: The readme is still referring to` stripe.PreviewAPIMode` though

File: api_version.go
Comment (xavdid-stripe) [reply api_version.go:10]: good call! Changed the base branch, so this will be a no-op. I'm not sure GH will reflect that though

File: README.md
Comment (jar-stripe) [inline README.md:591]: do users need to specify Content-Type for this to work w/ v2 and JSON?

File: rawrequest/client_test.go
Comment (jar-stripe) [inline rawrequest/client_test.go:40]: may want to assert on contentType == application/json, or return a 400 if its not, or something

File: rawrequest/client_test.go
Comment (jar-stripe) [inline rawrequest/client_test.go:81]: this isnt correct for a v1 api, right?  i.e. should be form encoded?

File: rawrequest/client_test.go
Comment (jar-stripe) [inline rawrequest/client_test.go:108]: this just returns empty string if it is not found; we should assert against it and fail if its not right.

File: stripe.go
Comment (jar-stripe) [inline stripe.go:495]: do we ever issue put or patch calls? if so, we want to keep the body in the body for those

File: stripe_test.go
Comment (jar-stripe) [inline stripe_test.go:1382]: adding this comment again: r.Header.Get (I _think_) will always "succeed" so we should have an assert or some kind of evaluation where we want to check that the header is of a specific value

File: README.md
Comment (jar-stripe) [reply README.md:591]: nm, answered my question as I went along; we use the uri prefix (v1 or v2) to determine content type, right?

File: stripe.go
Comment (jar-stripe) [reply stripe.go:495]: I think I answered my own question here; it looks like we restrict that above.  Given that, I am curious why this change was needed.  GETs are the only methods that cannot have a body.

**Review by jar-stripe — CHANGES_REQUESTED** [review]
Left a few non-blocking comments, but it looks like this repo is not set up to prevent merges on open conversations, so I wanted to not approve it just yet.

File: README.md
Comment (xavdid-stripe) [reply README.md:591]: correct, they can't specify API mode or content type manually. there didn't seem to be a point if we could just do it for them

File: rawrequest/client_test.go
Comment (xavdid-stripe) [reply rawrequest/client_test.go:40]: we do farther down - this line stores the incoming value on the "server" and then we assert its value after the response finishes.

File: rawrequest/client_test.go
Comment (xavdid-stripe) [reply rawrequest/client_test.go:81]: the test is technically wrong, yeah. this would be a 400 on the real API, but the tests treat `content` as opaque strings. The important part is that content-type is set correctly and the response is decoded. 

That said, I can update if you'd like!

File: stripe.go
Comment (xavdid-stripe) [reply stripe.go:495]: That was copied from the original PR: https://github.com/stripe/stripe-go/pull/1648/files#diff-f34528b48beca228ad5851464b3e20496f8c34c1f7b663e7f109fa9f5ed02fc4R426

I guess richard wanted the behavior to apply to `DELETE` in addition to `GET`? I don't have strong feelings, but I mostly erred towards not changing what's been in the beta branch 🤷

File: rawrequest/client_test.go
Comment (xavdid-stripe) [reply rawrequest/client_test.go:40]: https://github.com/stripe/stripe-go/pull/1924/files#diff-d25e28f49be803aa8fb79967e7cacada37a257743d0d82e8c4585b5a26ab9e1fR56

Comment (jar-stripe) [reply rawrequest/client_test.go:40]: ah i see that now.  thanks!

File: rawrequest/client_test.go
Comment (jar-stripe) [reply rawrequest/client_test.go:81]: i think it makes sense to update just from a readability standpoint; this might confuse me if I were new or unfamiliar.

File: stripe.go
Comment (jar-stripe) [reply stripe.go:495]: thats fair. can you update the comment?

**Comment by xavdid-stripe** [comment]
@jar-stripe updated the comment, test string body, and extra HTTP method validation in https://github.com/stripe/stripe-go/pull/1924/commits/a05ed2d66ee340ab603a045e3de9e98f2a522154! I noticed that `CallRaw` was actually a different call path and could use validation as well.

**Review by jar-stripe — APPROVED** [review]
looks good!

**Pull request merged**
